TRT-2017: Port regression tracking from bigquery to postgresql

dgoodwin opened the pull request:
This change will allow us to begin introducing triage / jira assignments without the snapshot the universe hacks and transactional issues that arise due to bigquerys inability to update rows in the streaming buffer for ~90 minutes. My hope is our future direction will be bigquery for big data, postgres for small data that requires updates. 

Component readiness does not require postgres today, this remains the case with this pr, but regression tracking will be disabled in that case. Going forward, we are looking more and more tied.

Includes migration, when the regression tracking command runs it checks if no entries exist in postgres, and assumes this is an indicator to transfer over from bigquery. I'll remove that code in a future PR.

This PR also includes some additional testing on the postgresql portion via the e2e framework we seldom use.

dgoodwin commented on hack/go-lint.sh:
```diff
@@ -28,6 +28,6 @@ else
     --volume "${PWD}:/go/src/github.com/openshift/sippy${VOLUME_OPTION}" \
     --workdir /go/src/github.com/openshift/sippy \
-    docker.io/golangci/golangci-lint:v1.56.2 \
+    docker.io/golangci/golangci-lint:v1.57.0 \
```
This lets make lint work locally as it now is compatible with our go mod version. However it picks up a couple new issues we're not addressing yet. I'll try to get this in a future PR.

stbenjam replied:
Which is what CI uses but only after merge

dgoodwin replied:
This implies fixing the lint warnings in this PR, which is probably ok, I'll do that now.

xueqzhan commented on pkg/api/componentreadiness/regressiontracker.go:
```diff
@@ -116,83 +176,78 @@ type RegressionTracker struct {
 // Run iterates all views with regression tracking enabled and syncs the results of it's
 // component report to the regression tables in bigquery.
```
Not bigquery any more

xueqzhan reviewed (commented):
Looks simpler than the BQ way!

dgoodwin commented on pkg/api/componentreadiness/component_report.go:
```diff
@@ -269,14 +273,17 @@ func (c *ComponentReportGenerator) GenerateReport(ctx context.Context) (crtype.C
-		log.WithError(err).Error("error listing current regressions")
-		errs = append(errs, err)
-		return crtype.ComponentReport{}, errs
+	// We only execute this if we were given a postgres database connection, it is still possible to run
```
I was not able to maintain dual compatible implementations. There was an older bigquery implementation of the interface I brought back because it supported updating records, but we had to throw it away because of the streaming buffer and implement the snapshot the world approach. This resulted in a different interface that doesn't really match how we would work with postgresql. 

I guess perhaps we could do an interface at the RegressionStore level itself with just a very high level "Run", but we'd have an unused impl laying around with no tests which doesn't feel great either. We could always bring it back to life via git if a use case materializes though.

xueqzhan replied:
I thought the same. Just checking. All frequent update tables should probably be postgres.

dgoodwin commented:
Rebased, addressed comment, and spotted a suspiciously unused param in FindOpenRegression, looks like a bug, I've addressed in separate commit.

dgoodwin commented:
It does indeed look so much simpler than bq was, quite funny going through it realizing the hoops we were jumping through.

stbenjam commented on test/e2e/componentreadiness/regressiontracker/regressiontracker_test.go:
```diff
@@ -0,0 +1,178 @@
+	require.NotEqual(t, "", os.Getenv("SIPPY_E2E_DSN"),
+		"SIPPY_E2E_DSN environment variable not set")
+
+	psqlFlags := flags.NewPostgresDatabaseFlags(os.Getenv("SIPPY_E2E_DSN"))
```
What's the reason behind using a different environment variable here? It'd be better if NewPostgresDatabaseFlags didn't take arguments.

dgoodwin replied:
Maybe I can find a way to set the flag opt programatically without modifying the constructor.

dgoodwin replied:
A simple flags.DSN did the trick, new change pushed

dgoodwin replied:
That's a good point, it's no code reuse at all really, I'll flip it to no use of flags.

stbenjam commented on scripts/e2e.sh:
```diff
@@ -33,24 +38,25 @@ $DOCKER run --name $PSQL_CONTAINER -e POSTGRES_PASSWORD=password -p $PSQL_PORT:5
 echo "Wait 5s for postgresql to start..."
 sleep 5
 
+export SIPPY_E2E_DSN="postgresql://postgres:password@localhost:$PSQL_PORT/postgres"
```
Why can't this just be SIPPY_DATABSE_DSN?

dgoodwin replied:
This e2e cleans out the regressions table after each test. I wanted to be exceedingly careful someone doesn't accidentally run it against prod, so I thought it made sense to make it a very explicit e2e var that isn't going to accidentally be set in any situation I could come up with. WDYT?

dgoodwin commented:
updated, the lint version now in sync with build root, addressed a few lint fixes that become required on new version, no longer using dbflags for postgres.

stbenjam commented:
/lgtm
